File: examples/typescript/researcher.ts
import { env } from "node:process";

import { createOpenAI } from "@ai-sdk/openai";
import { generateText, jsonSchema, stepCountIs, tool } from "ai";
import { Agent, defineTool, rebunoFetch, type RebunoTool } from "rebuno";

const SYSTEM_PROMPT =
  "You are a research assistant. You have access to tools for searching the web, " +
  "fetching documents, and doing math. Use them to answer the user's question. " +
  "When you have enough information, provide a clear final answer.";

const MODEL = env.OPENAI_MODEL ?? "gpt-4o-mini";

const webSearch = defineTool({
  name: "web_search",
  description: "Search the web for information about a topic.",
  inputSchema: { type: "object", properties: { query: { type: "string" } }, required: ["query"] },
  execute: async ({ query }: { query: string }) => {
    console.log(`web_search: ${query}`);
    await sleep(200);
    return {
      query,
      results: [
        { title: `Result for: ${query}`, snippet: `Info about ${query}.` },
        { title: `More on: ${query}`, snippet: `Details about ${query}.` },
      ],
    };
  },
});

const docFetch = defineTool({
  name: "doc_fetch",
  description: "Fetch and read the contents of a document at a URL.",
  inputSchema: { type: "object", properties: { url: { type: "string" } }, required: ["url"] },
  execute: async ({ url }: { url: string }) => {
    console.log(`doc_fetch: ${url}`);
    await sleep(100);
    return { url, title: `Document from ${url}`, content: `Mock content fetched from ${url}.`, word_count: 42 };
  },
});

const calculator = defineTool({
  name: "calculator",
  description: "Evaluate a mathematical expression. Example: '2 + 3 * 4'.",
  inputSchema: { type: "object", properties: { expression: { type: "string" } }, required: ["expression"] },
  execute: async ({ expression }: { expression: string }) => {
    console.log(`calculator: ${expression}`);
    return { expression, result: safeEval(expression) };
  },
});

async function process(input: { query: string }): Promise<Record<string, unknown>> {
  const openai = createOpenAI({ fetch: rebunoFetch });
  const { text } = await generateText({
    model: openai(MODEL),
    system: SYSTEM_PROMPT,
    prompt: input.query,
    tools: {
      web_search: asAiTool(webSearch),
      doc_fetch: asAiTool(docFetch),
      calculator: asAiTool(calculator),
    },
    stopWhen: stepCountIs(10),
  });
  console.log(`done: ${text}`);
  return { query: input.query, answer: text };
}

/** Expose a durable Rebuno tool to the Vercel AI SDK; its execute routes through the kernel. */
function asAiTool(t: RebunoTool<any, any>) {
  return tool({
    description: t.description,
    inputSchema: jsonSchema(t.inputSchema as Record<string, unknown>),
    execute: (args) => t.execute(args as Record<string, unknown>),
  });
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

function safeEval(expr: string): number {
  if (!/^[\d+\-*/().\s]+$/.test(expr)) throw new Error(`Unsupported expression: ${expr}`);
  return Function(`"use strict"; return (${expr});`)() as number;
}

const agent = new Agent("researcher", {
  secret: env.REBUNO_AGENT_SECRET ?? "researcher-secret",
  kernelUrl: env.REBUNO_URL ?? "http://localhost:8080",
});
console.log("researcher agent listening on :5001");
await agent.serve({ port: Number(env.AGENT_PORT ?? 5001) }, process);

File: examples/python/researcher.py
import ast
import asyncio
import logging
import operator
import os

from langchain.agents import create_agent
from langchain_openai import ChatOpenAI

from rebuno import Agent, http_client, tool

logging.basicConfig(
    level=logging.INFO, format="%(asctime)s %(name)s %(levelname)s %(message)s"
)
logger = logging.getLogger("researcher")

SYSTEM_PROMPT = (
    "You are a research assistant. You have access to tools for searching the web, "
    "fetching documents, and doing math. Use them to answer the user's question. "
    "When you have enough information, provide a clear final answer."
)

MODEL = os.environ.get("OPENAI_MODEL", "gpt-4o-mini")


@tool("web_search")
async def web_search(query: str) -> dict:
    """Search the web for information about a topic."""
    logger.info("web_search: %s", query)
    await asyncio.sleep(0.2)
    return {
        "query": query,
        "results": [
            {"title": f"Result for: {query}", "snippet": f"Info about {query}."},
            {"title": f"More on: {query}", "snippet": f"Details about {query}."},
        ],
    }


@tool("doc_fetch")
async def doc_fetch(url: str) -> dict:
    """Fetch and read the contents of a document at a URL."""
    logger.info("doc_fetch: %s", url)
    await asyncio.sleep(0.1)
    return {
        "url": url,
        "title": f"Document from {url}",
        "content": f"Mock content fetched from {url}.",
        "word_count": 42,
    }


@tool("calculator")
async def calculator(expression: str) -> dict:
    """Evaluate a mathematical expression. Example: '2 + 3 * 4'."""
    logger.info("calculator: %s", expression)
    return {"expression": expression, "result": _safe_eval(expression)}


async def process(query: str) -> dict:
    llm = ChatOpenAI(model=MODEL, temperature=0, http_async_client=http_client())
    graph = create_agent(
        model=llm,
        tools=[web_search, doc_fetch, calculator],
        system_prompt=SYSTEM_PROMPT,
    )
    result = await graph.ainvoke({"messages": [{"role": "user", "content": query}]})
    answer = result["messages"][-1].content
    logger.info("done: %s", answer)
    return {"query": query, "answer": answer}


_OPS = {
    ast.Add: operator.add,
    ast.Sub: operator.sub,
    ast.Mult: operator.mul,
    ast.Div: operator.truediv,
    ast.USub: operator.neg,
    ast.UAdd: operator.pos,
}


def _safe_eval(expr: str) -> float:
    return _eval_node(ast.parse(expr, mode="eval").body)


def _eval_node(node: ast.AST) -> float:
    if isinstance(node, ast.Constant) and isinstance(node.value, (int, float)):
        return node.value
    if isinstance(node, ast.BinOp) and type(node.op) in _OPS:
        return _OPS[type(node.op)](_eval_node(node.left), _eval_node(node.right))
    if isinstance(node, ast.UnaryOp) and type(node.op) in _OPS:
        return _OPS[type(node.op)](_eval_node(node.operand))
    raise ValueError(f"Unsupported expression: {type(node).__name__}")


if __name__ == "__main__":
    agent = Agent(
        "researcher",
        secret=os.environ.get("REBUNO_AGENT_SECRET", "researcher-secret"),
        base_url=os.environ.get("REBUNO_URL", "http://localhost:8080"),
    )
    logger.info("researcher agent listening on :5001")
    agent.run(process, port=int(os.environ.get("AGENT_PORT", "5001")))
